add link to exercism CLI (#571)
exercism/bash: docs/TESTS.md

# Tests

You can use the online editor on the Exercism website to solve the exercises.
Or, download the exercises to your computer, solve them locally and then submit them to Exercism.
The rest of this document talks about working locally.

## Exercism CLI

> You'll need the `exercism` command-line tool to download and submit exerices.

Reviewer: ```suggestion
You will need the `exercism` command-line tool to download and submit exercises.
```

> See [the documentation for it][exercism-cli].

Reviewer: ```suggestion
See [the exercism cli documentation][exercism-cli].
```

> ### Wrapper function for exercism tool

Reviewer: ```suggestion
### Wrapper function for the exercism tool
```

One annoying aspect of the default exercism tool is that, after you download an exercise, you have to change directory manually.

> This wrapper function handles that for you: store this in your `~/.bashrc`.

Reviewer: ```suggestion
This wrapper function handles that for you; store this in your `~/.bashrc`.
```

```bash
exercism () {
    local out
    readarray -t out < <(command exercism "$@")
    printf '%s\n' "${out[@]}"
    if [[ $1 == "download" && -d "${out[-1]}" ]]; then
        cd "${out[-1]}" || return 1
    fi
}
```
Note that cannot be a shell _script_, it must be a shell _function_.

> Otherwise the effect of `cd` cannot be executed in your current interactive shell.

Reviewer: ```suggestion
Shell scripts are executed in a child process and cannot effect the environment (like the current working directory) of your current shell.
```

## Testing locally

As there isn't much of a bash ecosystem, there also isn't really a defacto leader in the bash testing area.
For these examples we are using [bats][bats].

Run the tests for the hypothetical "whatever" exercise like this:
```bash
cd /path/to/your/exercise_workspace/bash/whatever
bats whatever.bats
```

For help on the meaning of the various `assert*` commands in the tests, refer to the documentation for the [bats-assert][bats-assert] library.

## Installing `bats-core`

You should be able to install it from your favorite package manager:

### For Mac (brew)

> On OS X with [Homebrew][homebrew] this would look something like this:

Reviewer: Isn't this now "macOS"?

Author: Yes it is.

#### Other Linux

For other Linux distributions the implementation of `bats` we use is not conveniently packaged.

> The best way to install it is from source: if you want to install it under `/usr/local` then

Reviewer: ```suggestion
The best way to install it is from source.
If you want to install it under `/usr/local` then
```

```
Note: When you are using the outdated `https://github.com/sstephenson/bats.git` and you discover an error like `cp: cannot create symbolic link '${HOME}/bin/bats': No such file or directory`, you have to copy the `bin/bats/libexec/` folder content to `${HOME}/bin/` manually.

There are reports that [newer bats versions don't behave well on MinGW bash][mingw-issues] -- before you run the install script, you might want to:
```
$ git checkout v1.1.0
```

## Legacy `bats`

`bats-core` was forked from [the original `bats` implementation][sstephenson-bats].

> The sstephenson/bats was quite buggy and had been abandoned.

Reviewer: ```suggestion
The `sstephenson/bats` project was quite buggy and had been abandoned.
```

Ownership was handed over in 2017: [sstephenson/bats#150 (comment)][bats-fork].

> If you have the original sstephenson/bats installed (check with `bats -v` reporting a version number less than 1.0), then you should switch to bats-core: otherwise you may find yourself [experiencing unexplained test failures][legacy-failures].

Reviewer: ```suggestion
If you have the original sstephenson/bats installed (check with `bats -v` reporting a version number less than 1.0), then you should switch to bats-core; otherwise you may find yourself [experiencing unexplained test failures][legacy-failures].
```
